File: fastapi + react-native/Robot-Control-Simulator/backend/app/models/robot.py
from deps import Directions, Event
from typing import Literal

class robot:
    def __init__(self, workspace_width: int, workspace_height: int):
        self.workspace_width = workspace_width
        self.workspace_height = workspace_height
        self.pos_x = 0
        self.pos_y = 0
        self.dir = Directions.NORTH
        self.warning_event = Event()
        self.report_event = Event()
        self.start_moving = False
        self.output = list[str]
        
    def is_in_workspace(self,x: int, y: int) -> bool:
        if x >= 0 and x < self.workspace_width and y >= 0 and y < self.workspace_height:
            return True
        else:
            self.warning_event.invoke()
            return False
    
    def place(self, x: str, y: str, facing: str):
        new_x = int(x)
        new_y = int(y)
        if not self.is_in_workspace(new_x, new_y): return
        
        self.pos_x = new_x
        self.pos_y = new_y
        self.dir = Directions[facing]
        self.start_moving = True
        
    def move(self):
        if not self.start_moving: return
        
        new_x = self.pos_x
        new_y = self.pos_y
        
        if self.dir == Directions.NORTH:
            new_y += 1
        elif self.dir == Directions.EAST:
            new_x += 1
        elif self.dir == Directions.SOUTH:
            new_y -= 1
        elif self.dir == Directions.WEST:
            new_x -= 1
        
        if self.is_in_workspace(new_x, new_y):
            self.pos_x = new_x
            self.pos_y = new_y       
    
    def rotate(self, to: Literal["LEFT", "RIGHT"]):
        if not self.start_moving: return
        
        if to == "LEFT":
            self.dir = Directions((self.dir.value - 1) % 4)
        elif to == "RIGHT":
            self.dir = Directions((self.dir.value + 1) % 4)
            
    def report(self):
        if not self.start_moving: return
        self.report_event.invoke()

File: c sharp/ToyRobotSimulator/Models/Robot.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RobotControlSimulator.UserControls;

namespace RobotControlSimulator.Models;

public class Robot(Workspace ws)
{
    public int posX = -1;
    public int posY = -1;
    public Direction facing = Direction.NORTH;

    public bool startMove = false;
    public string message = string.Empty;

    public event EventHandler? eh_Falling;
    public event EventHandler? eh_Report;

    public bool IsInWorkspace(int x, int y)
    {
        if (x >= 0 && x < ws.width && y >= 0 && y < ws.height)
        {
            ws?.Danger(false);
            return true;
        }

        eh_Falling?.Invoke(this, EventArgs.Empty);
        ws?.Danger(true);
        return false;
    }

    public void Place(int x, int y, Direction dir)
    {
        // Make sure the robot is placed in the workplace
        if (!IsInWorkspace(x, y)) return;

        // Place the robot
        posX = x;
        posY = y;
        facing = dir;

        // Since the robot has been places, the following commands will be able to run
        startMove = true;

        ws?.Place(x, -y, facing);
    }

    public void Move()
    {
        if (!startMove) return;

        // Make sure the robot is placed in the workplace
        if (!IsInWorkspace(posX, posY)) return;

        // Record the current position
        int moveX = 0;
        int moveY = 0;

        // Move one unit to the facing facing
        // Origin(0,0) is the South West most corner
        switch (facing)
        {
            case Direction.NORTH:
                moveY = 1;
                break;
            case Direction.EAST:
                moveX = 1;
                break;
            case Direction.SOUTH:
                moveY = -1;
                break;
            case Direction.WEST:
                moveX = -1;
                break;
            default:
                break;
        }

        // If the robot will move out of the workspace, ignore this movement
        if (!IsInWorkspace(posX + moveX, posY + moveY)) return;

        // Otherwise, apply this movement
        posX += moveX;
        posY += moveY;

        ws?.Move(moveX, -moveY);
    }

    public void Rotate(string dir)
    {
        if (!startMove) return;

        // Make sure the robot is placed in the workplace
        if (!IsInWorkspace(posX, posY)) return;

        switch (dir)
        {
            case "LEFT":
                facing = (Direction)(((int)facing + 3) % 4);
                break;
            case "RIGHT":
                facing = (Direction)(((int)facing + 1) % 4);
                break;
            default:
                break;
        }

        ws?.Rotate(facing);
    }

    public void Report()
    {
        if (!startMove) return;

        message = $"{posX},{posY},{facing}";

        eh_Report?.Invoke(this, EventArgs.Empty);
    }

    public void Reset()
    {
        startMove = false;
    }
}
